File: java/interviewquestions/ListMth.java
package interviewquestions;

/**
 * Given an array of characters, convert it to a linked list and find the
 * mth-to-last of the list.
 */
public class ListMth extends TestCase {
  public static class LinkedList {
    public static class Node {
      public Node next;
      public int value;
    }

    public Node head = null;
    public boolean isEmpty() { return head == null; }
  }

  public static LinkedList.Node mthToLast(char[] arr, int mth) {
    LinkedList lst = convertToList(arr);

    // use two pointers, one that starts at the first item and one
    // that starts m items later so that when the latter ends up
    // being null, the former is the mth-to-last item
    LinkedList.Node mthToLast = lst.head;
    LinkedList.Node advanced = lst.head;

    int i = 0;
    while (i < mth && advanced != null) {
      advanced = advanced.next;
      i++;
    }

    if (i < mth) return null;

    while (advanced != null) {
      advanced = advanced.next;
      mthToLast = mthToLast.next;
    }

    return mthToLast;
  }

  private static LinkedList convertToList(char[] arr) {
    LinkedList lst = new LinkedList();
    LinkedList.Node tmp;
    int i;

    if (lst.isEmpty()) {
      i = 1;
      tmp = new LinkedList.Node();
      tmp.value = arr[0];
      lst.head = tmp;
    } else {
      i = 0;
      tmp = lst.head;
    }

    while (i < arr.length) {
      tmp.next = new LinkedList.Node();
      tmp.next.value = arr[i];
      tmp = tmp.next;
      i++;
    }

    return lst;
  }

  public static void main(String[] args) {
    assertEquals(mthToLast(new char[] { 'a', 'b', 'c', 'd' }, 4).value, 'a');
    assertEquals(mthToLast(new char[] { 'a', 'b', 'c', 'd' }, 3).value, 'b');
    assertEquals(mthToLast(new char[] { 'a', 'b', 'c', 'd' }, 2).value, 'c');
    assertEquals(mthToLast(new char[] { 'a', 'b', 'c', 'd' }, 1).value, 'd');
    assertEquals(mthToLast(new char[] { 'e', 'f', 'g', 'h' }, 2).value, 'g');
    assertNull(mthToLast(new char[] { 'e', 'f', 'g', 'h' }, 5));
    assertEquals(mthToLast(new char[] { 'a', 'b', 'c', 'd',
                                        'e', 'f', 'g', 'h',
                                        'i', 'j', 'k', 'l',
                                        'm', 'n', 'o', 'p',
                                        'q', 'r', 's', 't',
                                        'u', 'v', 'w', 'x',
                                        'y', 'z' }, 26).value, 'a');
    assertEquals(mthToLast(new char[] { 'a', 'b', 'c', 'd',
                                        'e', 'f', 'g', 'h',
                                        'i', 'j', 'k', 'l',
                                        'm', 'n', 'o', 'p',
                                        'q', 'r', 's', 't',
                                        'u', 'v', 'w', 'x',
                                        'y', 'z' }, 1).value, 'z');
  }
}

File: python/interviewquestions/list_mth.py
"""
Given an array of characters, convert it to a linked list and find the
mth-to-last of the list.
"""

import unittest
import itertools


def mth_to_last(lst, mth):
    return lst[::-1][mth - 1]


def mth_to_last_alternative(lst, mth):
    """
    Because the other one was way too easy and leveraged python way more than
    it should in a proper job interview
    """
    # use a second pointer to the list that is m items forward,
    # iterate both lists together and stop as soon as the second list
    # runs out of elements
    for i1, i2 in itertools.izip_longest(lst, lst[mth:]):
        # I guess even this one uses a bit too much python maybe?
        # anyway, izip_longest is needed to pad to the longest of the lists,
        # otherwise (the classic 'zip') will truncate the iteration at the shortest
        # of the two
        if i2 is None:
            return i1
    return None


class TestListMth(unittest.TestCase):
    def test_list_mth(self):
        chars = list("abcdefghijklmnopqrstuvwxyz")
        self.assertEqual(mth_to_last(chars[:4], 4), "a")
        self.assertEqual(mth_to_last(chars[:4], 3), "b")
        self.assertEqual(mth_to_last(chars[:4], 2), "c")
        self.assertEqual(mth_to_last(chars[:4], 1), "d")
        self.assertEqual(mth_to_last(chars, 26), "a")
        self.assertEqual(mth_to_last(chars, 1), "z")

    def test_list_mth_alternative(self):
        chars = list("abcdefghijklmnopqrstuvwxyz")
        self.assertEqual(mth_to_last_alternative(chars[:4], 4), "a")
        self.assertEqual(mth_to_last_alternative(chars[:4], 3), "b")
        self.assertEqual(mth_to_last_alternative(chars[:4], 2), "c")
        self.assertEqual(mth_to_last_alternative(chars[:4], 1), "d")
        self.assertEqual(mth_to_last_alternative(chars, 26), "a")
        self.assertEqual(mth_to_last_alternative(chars, 1), "z")


if __name__ == "__main__":
    unittest.main()
